broker/src/main/java/org/apache/rocketmq/broker/client/ConsumerGroupInfo.java

```diff
@@ -120,9 +121,9 @@ public boolean updateChannel(final ClientChannelInfo infoNew, ConsumeType consum
         this.messageModel = messageModel;
         this.consumeFromWhere = consumeFromWhere;
 
-        ClientChannelInfo infoOld = this.channelInfoTable.get(infoNew.getChannel());
+        ClientChannelInfo infoOld = this.channelInfoTable.get(infoNew.getRemotingChannel());
         if (null == infoOld) {
-            ClientChannelInfo prev = this.channelInfoTable.put(infoNew.getChannel(), infoNew);
+            ClientChannelInfo prev = this.channelInfoTable.put(infoNew.getRemotingChannel(), infoNew);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **124, 125, and 126**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread's `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and uses that reference at line: **129** and **133**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.